File: code/ch20/20.3.1. 1277.countSquares.py
class Solution:
    def countSquares(self, matrix: List[List[int]]) -> int:
        res = 0
        m = len(matrix)
        n = len(matrix[0])
        dp = [[0] * (n + 1) for _ in range(m + 1)]
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if matrix[i - 1][j - 1] == 1:
                    dp[i][j] = min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1
                res += dp[i][j]
        return res

File: code/ch20/20.3.1. 1277.countSquares.cpp
#include <vector>
#include <cstdlib>
#include <algorithm>
using namespace std;

class Solution {
public:
  int countSquares(vector<vector<int> >& matrix) {
    int res = 0, m = matrix.size(), n = matrix[0].size();
    int **dp = (int **)malloc((m + 1) * sizeof(int *));
    for (int i = 0; i <= m; i++) {
      dp[i] = (int *)malloc((n + 1) * sizeof(int));
      for (int j = 0; j <= n; j++) dp[i][j] = 0;
    }
    for (int i = 1; i <= m; i++)
      for (int j = 1; j <= n; j++) {
        if (matrix[i - 1][j - 1] == 1)
          dp[i][j] = min(min(dp[i - 1][j], dp[i][j - 1]),\
              dp[i - 1][j - 1]) + 1;
        res += dp[i][j];
      }
    for (int i = 0; i <= m; i++) free(dp[i]);
    free(dp);
    return res;
  }
};
